### DISCUSSION TITLE
(stepfunctions): Cannot make a loop with Choice and afterwards() in a state machine

### DISCUSSION THREAD

**Discussion by nebur395** [aws/aws-cdk#19718] [Q&A]
### The Question
I would want to use a Chain-Choice flow to make a loop within a step function. Besides that, I would like to be able to follow the main chain at root level, not at condition level. I've tried to use `afterwards` method since I thought this was for implemented following this purpose but it doesn't seem to work or I'm missing some implementation details. 

### Context/Background
I'm splitting the state machine in several constructs for readability and encapsulation reasons. Simplifying the use case it would look something like this (notice it's written in TS and it's just pseudo-code):

```typescript
// In a real scenario this would be a list of steps, not just one
const setupStep1Chain = () => Chain.start(new LambdaInvoke('Step 1'));

const waitUntilStep1HasFinished = () => {
  const checkStep1HasFinished = new LambdaInvoke('Check step 1 status');
  const waitAndCheck = new Wait('Wait X').next(checkStep1HasFinished);
  const step2 = new LambdaInvoke('Step 2');
  const step1Choice = new Choice('Has step 1 finished?');

  return Chain.start(checkStep1HasFinished)
    .next(step1Choice.when('No', waitAndCheck).afterwards())
    .next(step2);
};

const setupStateMachine = () =>  
  new StateMachine({
    definition: Chain
      // You can imagine all these methods, as constructs in different files or workspaces or repos 
      // instead a simple function in the same file. This is just for make it simpler
      .start(setupStep1Chain())
      .next(waitUntilStep1HasFinished())
      // ... (a list of more chainable next steps or chains)
      .next(new Succeed('Done'));
  });
```

This is how I understood the method `afterwards()` should look like. However if I follow this approach, I just get this error:
>  Error: State 'Check step 1 status' already has a next state

And if I remove the `.afterwards()` method, I would get this other error:
> Cannot add to chain: last state in chain (Has step 1 finished?) does not allow it

I tried different approaches in the Choice step such as using `.otherwise()` and intermediate `Pass` state... and a lot of this, but nothing seems to work. Of course, I can follow the flow in the conditional statement, something like this:

```typescript
// In a real scenario this would be a list of steps, not just one
const setupStep1Chain = () => Chain.start(new LambdaInvoke('Step 1'));

const waitUntilStep1HasFinished = (nextChain) => {
  const checkStep1HasFinished = new LambdaInvoke('Check step 1 status');
  const waitAndCheck = new Wait('Wait X').next(checkStep1HasFinished);
  const step2 = new LambdaInvoke('Step 2');
  const step1Choice = new Choice('Has step 1 finished?');

  return Chain.start(checkStep1HasFinished)
    .next(step1Choice
      .when('No', waitAndCheck)
      .otherwise(step2.next(nextChain)));
};

const setupStateMachine = () =>  
  new StateMachine({
    definition: Chain
      .start(setupStep1Chain())
      .next(waitUntilStep1HasFinished(
        Chain. // ... (a list of more chainable next steps or chains)
          .next(new Succeed('Done'));
      ))
  });
```

However, this looks for me a bit weird for some architectural patterns, since you have to make `waitUntilStep1HasFinished()` method aware of the next chainable logic and as the author (of the issue I'm going to link at the end) said, you really decrease the readability of the code defining the state machine, having to read it by inspecting each Choice.

Is there any workaround to make this? Am I missing something in the afterwards feature?

Please, find just below a visual representation of what I would like to achieve with the pseudo-code I've just posted

Thanks a lot! 

![State Machine Loop and Afterwards](https://user-images.githubusercontent.com/17260233/130637811-9fdfdb8a-1617-4925-99d0-216f746a3dee.png)

### Environment

  - **CDK CLI Version:** 1.118.0
  - **Module Version:** 1.118.0
  - **Node.js Version:** 16.5.0
  - **OS:** macOs Big Sur 11.5.2
  - **Language (Version):** typescript 4.3.5

### Other information
This issue is highly related with [this one](https://github.com/aws/aws-cdk/issues/8680), it was closed without a response though

**Comment by nebur395** [comment]
Hi there, any news so far? It's a bit blocker for us. Thanks a lot!

**Comment by NGL321** [comment]
Hey @nebur395,

Thanks for reaching out! I have done some digging into this problem and it seems at its core this is a confusion of `afterwards()` vs `otherwise()` in a state machine workflow.

`afterwards()` is used to represent the `then` component of a `if... else... then...` condition. This requires there to be an `if` AND an `else`. In your first provided code block you are using `afterwards()` as the `else` component when this is actually the job of `otherwise()`. Your second code block correctly utilizes otherwise and creates the correct state machine flow. Remember, that with `otherwise` the ASL generated will look something like:
``` json
"Step1": {
	"Next": "Choice",
	...
},
"Choice": {
	"Choices": [{
		"CONDITION": "RESULT",
		"Next": "Step1"
	}],
	"Default": "Step2"
},
"Step2": {
	...
}
```

whereas ASL with `afterwards()` doesnt change the `Default` state, but rather injects an option after all choices are made. This is considered an extra step and runs regardless of the condition. This is useful for breaking out of complex loops done with conditions that have many possible results but all need to end with the same next step. Chances are, if you dont have an `otherwise()` already, you will not use an `afterwards()`.

Please let me know if you need more info on this!

😸 😷

**Comment by nebur395** [comment]
Good morning!

Thanks for the reply! :) Maybe I didn't make myself clear.  I understand `.otherwise()` would be kinda `else` statement on a conditional flow. But what I would like to do would be something like
```
...
if (whatever) {}
else {}
...continue here regardless the if-else clause 
```
 I don't have in this use case an else clause, (which would be the `.otherwise()` method), that's the reason I didn't include it in the example before, but I don't mind to include it. Then, the pseudo-code would be something like:
 ```typescript
 // In a real scenario this would be a list of steps, not just one
const setupStep1Chain = () => Chain.start(new LambdaInvoke('Step 1'));

const waitUntilStep1HasFinished = () => {
  const checkStep1HasFinished = new LambdaInvoke('Check step 1 status');
  const waitAndCheck = new Wait('Wait X').next(checkStep1HasFinished);
  const step2 = new LambdaInvoke('Step 2');
  const step3 = new LambdaInvoke('Step 3');
  const step1Choice = new Choice('Has step 1 finished?');

  return Chain.start(checkStep1HasFinished)
    .next(
      step1Choice.
        when('No', waitAndCheck)
        .otherwise(step2)
    );
};

const setupStateMachine = () =>  
  new StateMachine({
    definition: Chain
      // You can imagine all these methods, as constructs in different files or workspaces or repos 
      // instead a simple function in the same file. This is just for make it simpler
      .start(setupStep1Chain())
      .next(waitUntilStep1HasFinished())
      // ... (a list of more chainable next steps or chains)
      .next(new Succeed('Done'));
  });
 ```
 This will throw this error: `Cannot add to chain: last state in chain (Has step 1 finished?) does not allow it` because I'm adding a step (`new Succeed('Done')`) just after a choice.
 If I use a combination of `otherwise` and `afterwards` that would be something like:
 ```typescript
 // In a real scenario this would be a list of steps, not just one
const setupStep1Chain = () => Chain.start(new LambdaInvoke('Step 1'));

const waitUntilStep1HasFinished = () => {
  const checkStep1HasFinished = new LambdaInvoke('Check step 1 status');
  const waitAndCheck = new Wait('Wait X').next(checkStep1HasFinished);
  const step2 = new LambdaInvoke('Step 2');
  const step3 = new LambdaInvoke('Step 3');
  const step1Choice = new Choice('Has step 1 finished?');

  return Chain.start(checkStep1HasFinished)
    .next(
      step1Choice.
        when('No', waitAndCheck)
        .otherwise(step2)
        .afterwards()
    );
};

const setupStateMachine = () =>  
  new StateMachine({
    definition: Chain
      // You can imagine all these methods, as constructs in different files or workspaces or repos 
      // instead a simple function in the same file. This is just for make it simpler
      .start(setupStep1Chain())
      .next(waitUntilStep1HasFinished())
      // ... (a list of more chainable next steps or chains)
      .next(new Succeed('Done'));
  });
 ```
 This should be what I was looking for I guess. In order to follow the flow out of the conditional statements. However, since one of the flows is a loop, I will get this error: `Error: State 'Check step 1 status' already has a next state`. Which makes sanes because indeed, step 1 has already a next step, because it's a loop. 
 
 So I wonder if there is any possibility to give the flow control back to the main thread, cause I would wanna reach something similar to this:
 
 ```typescript
const setupStep1Chain = () => Chain.start(new LambdaInvoke('Step 1'));

const waitUntilStep1HasFinished = () => {
  const checkStep1HasFinished = new LambdaInvoke('Check step 1 status');
  const waitAndCheck = new Wait('Wait X').next(checkStep1HasFinished);
  const step1Choice = new Choice('Has step 1 finished?');

  return Chain.start(checkStep1HasFinished)
    .next(step1Choice.when('No', waitAndCheck));
};

const setupStateMachine = () =>  
  new StateMachine({
    definition: Chain
      .start(setupStep1Chain())
      .next(waitUntilStep1HasFinished())
      // ... (and I would want to continue the flow here!! :octocat: 
      .next(new Succeed('Done'));
  });
 ```
 
 I hope I made myself clearer now :)

**Comment by m-radzikowski** [comment]
I found this issue because of the same problem.

I understand `.afterwards()` as a way to flatten the chain definition and not make it nested every time you put Choice in there. But it doesn't work this way, as presented by @nebur395.

What I expected, using the example from above:

```ts
Chain
    .start(setupStep1Chain())
    .next(waitUntilStep1HasFinished())
    .next(new Pass('Done'));
```

what I have to do:

```ts
Chain
    .start(setupStep1Chain())
    .next(waitUntilStep1HasFinished(new Chain
      .start(new Pass('Done'))
      .next(...)
    ))

const waitUntilStep1HasFinished = (nextStep) => {
    const wait = new Wait()
    return Chain
        .start(wait)
        .next(
            new Choice()
                .when(Condition.something(), wait)
                .otherwise(nextStep)
        )
}
```

so each Chain is making a nested structure, while it could be flattened out.

**Comment by watanabeshuji** [comment]
I have same problem.

I found Parallel is effective for this case.
(I don't think best solution...)

```
    const definition = submitJob
      .next(
        new sfn.Parallel(scope, 'Parallel')
        .branch(
          waitX
          .next(getStatus)
          .next(new sfn.Choice(this, 'Job Complete?')
            // Look at the "status" field
            .when(sfn.Condition.stringEquals('$.status', 'FAILED'), jobFailed)
            .when(sfn.Condition.stringEquals('$.status', 'SUCCEEDED'), finalStatus)
            .otherwise(waitX)
          )
        )
      )
      .next(new sfn.Pass(this, 'Done'))
```
![スクリーンショット 2022-08-22 18 54 01](https://user-images.githubusercontent.com/5770822/185893813-4c107b23-9544-4ae4-b903-416414df158e.png)

**Reply by vesko-vujovic** [reply]
Is there a way to track the number of attempts? to not be stuck inside this loop of waiting for the job to complete.

Is there any idea?

**Comment by enpatrik** [comment]
I had the same problem, managed to solve it thanks to previous comments and by using `sfn.Chain.sequence` or `sfn.Chain.custom`.

Example:

```typescript
const step1 = new sfn.Pass(this, 'step1')
const step2 = new sfn.Pass(this, 'step2')

const step1Choice = new sfn.Choice(this, 'check?')
  .when(sfn.Condition.booleanEquals('$.foo', true), step2, { comment: 'finished!' })
  .otherwise(new sfn.Wait(this, 'wait', { time: sfn.WaitTime.duration(cdk.Duration.seconds(15)) }).next(step1))
  .afterwards()
//const step1Chain = sfn.Chain.start(step1).next(step1Choice) // Instead of this, do:
const step1Chain = sfn.Chain.sequence(step1.next(step1Choice), step2)

const done = new sfn.Pass(this, 'done')

const chain = step1Chain.next(done)
```

![Screenshot 2024-06-27 at 4 27 46 PM](https://github.com/aws/aws-cdk/assets/1154268/11f95e15-6ba7-466b-874c-dd4173e92858)

**Comment by cscanlin** [comment]
Here's a working example of an entire chain in python that might help make things more clear

```
wait_state = sfn.Wait(
    self,
    'Wait 60 Seconds',
    time=sfn.WaitTime.duration(Duration.seconds(60))
)

chain = (
    sfn.Chain.start(self.tasks['submitter'])
    .next(self.tasks['monitor'])
    .next(
        sfn.Choice(self, 'Check Jobs Completed')
        .when(
            sfn.Condition.boolean_equals('$.jobs_completed', False),
            # This causes a loop back to monitor until the job is complete
            wait_state.next(self.tasks['monitor']),
        )
        # We can't use afterwards() here because the when condition
        # has a separate path through the chain.
        # So the remaining steps must be nested under the otherwise().
        .otherwise(
            sfn.Pass(self, 'Jobs Complete')
            .next(self.tasks['check_num_sites'])
            .next(
                sfn.Choice(self, 'Check Number of Sites Choice')
                .when(
                    sfn.Condition.number_greater_than('$.num_sites', 1000),
                    self.tasks['postprocessor_L'],
                )
                .when(
                    sfn.Condition.number_greater_than('$.num_sites', 100),
                    self.tasks['postprocessor_M'],
                )
                .otherwise(self.tasks['postprocessor_S'])
                .afterwards()
            )
            .next(sfn.Pass(self, 'Postprocessor Complete'))
        )
    )
)
```

Everything in the `otherwise()` condition must be in that deeper nested level in this case.

If we tried to put an `afterwords()`, and move all of the code after that (at the flattest level), we would see an issue like `monitor already has a next state`. You can imagine that the `afterwards()` puts a "hidden `next()` step" on both the code in the `when()` clause, and the code in the `otherwise()`. But that wouldn't work because the last state in the `when()` clause (`monitor`) already has a `next()` clause, which is the choice itself.

However in the postpressor below, we can use `afterwards()` because none of the other choices create a loop. So we can force them to all join back together with the "hidden `next()` step" of `afterwards()`, and then subsequent steps can be at the same level.
